### PR TITLE
Include BDBag TSV file generation in integration (#565)

### PR DESCRIPTION
**Pull request by mjkrause**
* checks BDBag endpoint
* asserts row count of each BDBag TSV file and of manifest to be > 1
* query parameter for filter set more stringent if `dss_query_prefix`
is not set
* uses no filter in query for BDBag endpoint test using personal
deployment
* uses less stringent filter for manifest endpoint test when using
personal deployment
* when using `dev` deployment filter constrains ES search results to
brain as organ for both, manifest and BDBag, in order to fit execution
time into lambda life cycle
* skips ES hit for BDBag files if dictionary containing ES search hits
does not contain key `specimens` or if the list of that key is empty

### CODE REVIEW

File: test/local_integration_test.py [hunk 1]
```diff
@@ -43,7 +47,7 @@ def tearDown(self):
     def test_webservice_and_indexer(self):
         test_uuid = str(uuid.uuid4())
         test_name = f'integration-test_{test_uuid}_{self.bundle_uuid_prefix}'
-        logger.info('Starting test using test name, %s ...', test_name)
```
Comment (mjkrause) [inline test/local_integration_test.py]: Oops, was an accident, will revert.

**Comment by mjkrause** [comment]
Changes in the last commit allow use of `{'file': {}}` as the manifest filter for both, the manifest and the BDBag on the personal deployment. That simplifies the flow control in that test.

**Comment by mjkrause** [comment]
As discussed, I added a check of the single-part manifest export endpoint, and reverted code in class `ManifestResponse`, def `_extract_fields` back to the version of branch `develop`. 

Error message from log at https://console.aws.amazon.com/cloudwatch/home?region=us-east-1 logStream:group=/aws/lambda/azul-service-mjkrause, PST timestamp 8:34:50 h:
```
[DEBUG]	2019-03-20T15:34:50.56Z	8f9e9ffe-68a1-403b-89fb-1ce09c12cec4	Caught exception for <function get_manifest at 0x7f766f7f2510>
Traceback (most recent call last):
File "/var/task/chalice/app.py", line 726, in _get_view_function_response
response = view_function(**function_args)
File "/var/task/app.py", line 432, in get_manifest
response = es_td.transform_manifest(filters=filters, format=format)
File "/var/task/azul/service/responseobjects/elastic_request_builder.py", line 557, in transform_manifest
return manifest.return_response()
File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 392, in return_response
object_key = self._push_content()
File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 249, in _push_content
self._iterate_hit_tsv(hit, writer)
File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 350, in _iterate_hit_tsv
inner_entity_fields += self._extract_fields(entities, column_mapping)
File "/var/task/azul/service/responseobjects/hca_response_v5.py", line 225, in _extract_fields
field_value = entity[field_name]
KeyError: 'library_construction_approach'
```
This is what was expected based on my observations yesterday.

File: test/local_integration_test.py [hunk 2]
```diff
@@ -59,8 +63,14 @@ def test_webservice_and_indexer(self):
         self.check_endpoint_is_working(config.service_endpoint(), '/version')
         self.check_endpoint_is_working(config.service_endpoint(), '/repository/summary')
         self.check_endpoint_is_working(config.service_endpoint(), '/repository/files/order')
-        manifest_filter = {"file": {"organPart": {"is": ["temporal lobe"]}, "fileFormat": {"is": ["bai"]}}}
+        if config.dss_query_prefix is not '':  # only subset of indexed metadata, use less stringent filter
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]:
```suggestion
        if config.dss_query_prefix:  # only subset of indexed metadata, use less restrictive filter
```

You shouldn't compare strings with `is`. It works for empty and single-character strings but that isn't documented.

http://guilload.com/python-string-interning/

File: test/local_integration_test.py [hunk 3]
```diff
@@ -77,6 +87,30 @@ def check_endpoint_is_working(self, lambda_endpoint: str, url: str):
         url = lambda_endpoint + url
         response = self.requests.get(url)
         response.raise_for_status()
+        # Assert that manifest contains at least one row of metadata.
+        if '/manifest/files' in url:
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: This is ugly and only covers `/manifest/files` not `/repository/files/export`. Can you extract this into a separate function: `check_manifest(check_endpoint_is_working(…))`. For that to work, check_endpoint_is_working needs to return `response.content`.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 4]
```diff
@@ -343,9 +347,9 @@ def _iterate_hit_tsv(self, es_search_hit, writer):
 
     def _iterate_hit_bdbag(self, es_search_hit, participants: MutableSet[str], sample_writer):
         hit_dict = es_search_hit.to_dict()
-
-        assert len(hit_dict['contents']['specimens']) == 1
-        assert len(hit_dict['contents']['cell_suspensions']) == 1
+        # Some files are not associated with specimens (e.g., PDFs) - skip them.
+        if 'specimens' not in hit_dict['contents'].keys() or len(hit_dict['contents']['specimens']) == 0:
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply src/azul/service/responseobjects/hca_response_v5.py]: (I assumed this comment and the one above are one comment combined?)
When I ran the code I only saw `KeyError`, but in the index I saw PDFs which had a key `specimens` but it was an empty list. That's why I set both conditions. My query on `azul-index-dev` was
```
GET azul_files_dev/_search
{
  "query": {
    "match": {
      "contents.files.file_format": "pdf"
    }
  }
}
```

File: test/local_integration_test.py [hunk 5]
```diff
@@ -77,6 +87,30 @@ def check_endpoint_is_working(self, lambda_endpoint: str, url: str):
...
+    def check_bdbag_endpoint(self, lambda_endpoint: str, url: str):
+        url = lambda_endpoint + url
+        response = self.requests.get(url)
+        response.raise_for_status()
+        with ZipFile(BytesIO(response.content)) as zip_fh:
+            data_path = os.path.join(os.path.dirname(first(zip_fh.namelist())), 'data')
+            participant_file, sample_file = (
+                (os.path.join(data_path, filename) for filename in ('participant.tsv', 'sample.tsv'))
+            )
+            tsv_files = {os.path.basename(participant_file): participant_file,
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: Yes, true. Fixed.

File: test/local_integration_test.py (hunk 2 above)
Comment (mjkrause) [reply test/local_integration_test.py]: Fixed, not sure what I was thinking...

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjkrause) [reply src/azul/service/responseobjects/hca_response_v5.py]: Yet, if I remove the second condition the integration test passes on both, my personal and the dev deployment. So it's removed.

File: test/local_integration_test.py (hunk 3 above)
Comment (mjkrause) [reply test/local_integration_test.py]: Fixed.

File: test/local_integration_test.py [hunk 6]
```diff
@@ -77,6 +87,30 @@ def check_endpoint_is_working(self, lambda_endpoint: str, url: str):
...
+            self.assertTrue(num_rows > 1)
+            logger.info(f'Manifest contains {num_rows} rows.')
+
+    def check_bdbag_endpoint(self, lambda_endpoint: str, url: str):
+        url = lambda_endpoint + url
+        response = self.requests.get(url)
+        response.raise_for_status()
+        with ZipFile(BytesIO(response.content)) as zip_fh:
+            data_path = os.path.join(os.path.dirname(first(zip_fh.namelist())), 'data')
+            participant_file, sample_file = (
```
[the remark this answers is not in the document]
Comment (mjkrause) [reply test/local_integration_test.py]: This has been fixed using a dictionary comprehension, like you suggested in the comment right below.

**Comment by mjkrause** [comment]
I addressed all of your comments in the revised code, but unfortunately hadn't responded to them here in the communication. I just added my responses.

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/service/responseobjects/hca_response_v5.py]: The above query hits the contributions index for files. The manifest and bdbag code hits the aggregate index for files.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 7]
```diff
@@ -343,9 +347,9 @@ def _iterate_hit_tsv(self, es_search_hit, writer):
...
+            return
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/service/responseobjects/hca_response_v5.py]: Also check for presence of `cell_suspension`

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 8]
```diff
@@ -343,9 +347,9 @@ def _iterate_hit_tsv(self, es_search_hit, writer):
 
     def _iterate_hit_bdbag(self, es_search_hit, participants: MutableSet[str], sample_writer):
         hit_dict = es_search_hit.to_dict()
-
-        assert len(hit_dict['contents']['specimens']) == 1
-        assert len(hit_dict['contents']['cell_suspensions']) == 1
+        # Some files are not associated with specimens (e.g., PDFs) - skip them.
+        if 'specimens' not in hit_dict['contents'].keys():
+            return
         assert len(hit_dict['contents']['files']) == 1
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: Remove. The one() call takes care of this.

File: test/local_integration_test.py [hunk 9]
```diff
@@ -59,8 +63,17 @@ def test_webservice_and_indexer(self):
...
         self.check_endpoint_is_working(config.service_endpoint(), '/repository/summary')
         self.check_endpoint_is_working(config.service_endpoint(), '/repository/files/order')
-        manifest_filter = {"file": {"organPart": {"is": ["temporal lobe"]}, "fileFormat": {"is": ["bai"]}}}
-        self.check_endpoint_is_working(config.service_endpoint(), f'/manifest/files?filters={manifest_filter}')
+        if config.dss_query_prefix:  # only subset of indexed metadata, use less stringent filter
+            manifest_filter = {"file": {}}
+        else:
+            manifest_filter = {"file": {"organ": {"is": ["Brain"]}, "fileFormat": {"is": ["bai"]}}}
+        self.check_manifest(self.check_endpoint_is_working(config.service_endpoint(),
+                                                      f'/repository/files/export?filters={manifest_filter}'))
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: bad indent

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjkrause) [reply src/azul/service/responseobjects/hca_response_v5.py]: Changed it.

File: src/azul/service/responseobjects/hca_response_v5.py (hunk 8 above)
Comment (mjkrause) [reply src/azul/service/responseobjects/hca_response_v5.py]: Removed.

File: test/local_integration_test.py (hunk 9 above)
Comment (mjkrause) [reply test/local_integration_test.py]: Fixed.

File: src/azul/service/responseobjects/hca_response_v5.py [hunk 10]
```diff
@@ -343,37 +347,34 @@ def _iterate_hit_tsv(self, es_search_hit, writer):
...
-                                         specimen_fields[1],
-                                         cell_suspension_fields[0],
-                                         self._translate(bundle, 'bundles'),
-                                         file_fields,
-                                         [dss_url],
-                                         [drs_url]))
-            participant_id = specimen_fields[1][0]
-            participants.add(participant_id)
+        # Some files are not associated with specimens or cell suspensions (e.g., PDFs, JPEGs) - skip them.
+        if 'specimens' in hit_dict['contents'].keys() and 'cell_suspensions' in hit_dict['contents'].keys():
```
Comment (hannes-ucsc) [inline src/azul/service/responseobjects/hca_response_v5.py]: `k in d` is equivalent to `k in d.keys()`, but shorter

File: test/local_integration_test.py [hunk 11]
```diff
@@ -77,6 +90,28 @@ def check_endpoint_is_working(self, lambda_endpoint: str, url: str):
         url = lambda_endpoint + url
         response = self.requests.get(url)
         response.raise_for_status()
+        return response
+
+    def check_manifest(self, response):
+        """Assert that manifest contains at least one row of metadata."""
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: Triple quotes on separate line, please.

**Pull request merged**